src/App.tsx

```diff
@@ -1,223 +1,218 @@
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [tempNewTodo, setTempNewTodo] = useState<Todo | null>(null);
+  const [completedFilter, setCompletedFilter] = useState<Filter>(Filter.all);
+  const [processingTodoIds, setProcessingTodoId] = useState<number[]>([]);
+  const [isToggleAll, setIsToggleAll] = useState(false);
+  const [isAddingTodo, setIsAddingTodo] = useState(false);
+
   const user = useContext(AuthContext);
-  const newTodoField = useRef<HTMLInputElement>(null);
+  const [showError, closeError, errorMessages] = useError();
+
+  const filterTodos = useMemo(() => {
+    return getFilterTodos(todos, completedFilter);
+  }, [todos, completedFilter]);
+
+  const onAddTodo = useCallback(async (fieldsForCreate: Omit<Todo, 'id'>) => {
+    try {
+      setIsAddingTodo(true);
+      setTempNewTodo({
+        ...fieldsForCreate,
+        id: 0,
+      });
+
+      const newTodo = await addTodo(fieldsForCreate);
+
+      setTodos(prevState => [...prevState, newTodo]);
+    } catch {
+      showError('Unable to add todo');
+
+      throw Error('Error while adding todo');
+    } finally {
+      setIsAddingTodo(false);
+      setTempNewTodo(null);
+    }
+  },
+  [showError]);
+
+  const deleteTodoClick = useCallback(
+    (id: number) => {
+      setProcessingTodoId([id]);
+
+      deleteTodo(id)
+        .then(() => (
+          setTodos(currentTodos => currentTodos
+            .filter(todo => todo.id !== id))
+        ))
+        .catch(() => {
+          showError('Unable to delete a todo');
+        })
+        .finally(() => setProcessingTodoId([]));
+    },
+    [showError],
+  );
+
+  const deleteTodoCompleted = useCallback(
+    () => {
+      const completedTodoId = todos
+        .filter(todo => todo.completed)
+        .map(todo => todo.id);
+
+      setProcessingTodoId(completedTodoId);
+
+      const deletePromises = todos.map(todo => {
+        if (todo.completed) {
+          return deleteTodo(todo.id)
+            .then(() => setTodos(prev => prev
+              .filter(prevTodo => !completedTodoId.includes(prevTodo.id))));
+        }
+
+        return Promise.resolve();
+      });
+
+      Promise.all(deletePromises)
+        .finally(() => setProcessingTodoId([]));
+    },
+    [todos],
+  );
+
+  const changeFilterStatus = useCallback(
+    async (id: number, status: boolean) => {
+      setProcessingTodoId([id]);
+
+      try {
+        await updateTodoOnServer(id, { completed: status });
+        setTodos(prev => prev.map(todo => {
+          return todo.id === id
+            ? { ...todo, completed: status }
+            : todo;
+        }));
+      } catch (e) {
+        showError('Unable to update a todo');
+      } finally {
+        setProcessingTodoId([0]);
+      }
+    },
+    [showError],
+  );
+
+  const toggleAll = async () => {
+    const completedTodoId = todos
+      .filter(todo => todo.completed === isToggleAll)
+      .map(todo => todo.id);
+
+    setProcessingTodoId(completedTodoId);
```

Comment:
Why do you need this? `update` function already done this logic

Reply:
fix